### DISCUSSION TITLE
USB device redirection

### DISCUSSION THREAD

**Discussion by dsanttos12** [ClassicOldSong/Apollo#70] [Ideas]
Would it be possible to add support for USB device redirection, as occurs in virtualhere and usb network gate, so that sunshine/apollo users could use steering wheels, dongles, razer synapse and could use dualsense's adaptive triggers?☻

**Comment by ClassicOldSong** [comment]
It is planned but remote USB is a very tedious and complicated work to do. That is why VirtualHere and other remote USB solutions are so expensive.

As for now, it will not be implemented in the short term.

**Reply by dsanttos12** [reply]
How long will it take you to add the USB device redirection implementation?

**Reply by ClassicOldSong** [reply]
No guarantee, this is a free project and I cannot spare enough time for digging into this feature before I can finish my other projects that fill my financial needs.

**Reply by dsanttos12** [reply]
Would it be possible for me to help or make some contribution to the implementation of USB device redirection on artemis/apollo using existing open source libraries such as USB/IP?

**Reply by ClassicOldSong** [reply]
I've already tried USB/IP impl for Windows, It's still buggy and not suitable even for mouse passthrough, the mouse resolution is lower than the current emulated one and there're some major compatibility issues.

The most important thing is, USB/IP driver is a kernel mode driver which requires a purchased license, self-signed ones won't work, which isn't cheap.

**Reply by dsanttos12** [reply]
I saw on a github page that virtalhere is a proprietary version of USB/IP on this https://github.com/milaq/steamlink_VirtualHereFreeInstaller page in this following excerpt "VirtualHere is a proprietary version of the open source USB/IP project."

**Reply by ClassicOldSong** [reply]
No evidence it's a fork of USB/IP, although they should all be the same from principal level.

Even if it is, it's still heavily modified for compatibility with more devices.

The Steam version of virtual here and its unlimited license isn't that expensive compared to its standalone version, I think it's worth considering for the effort the creator has already put into it.

**Reply by dsanttos12** [reply]
The Steam license is not worth it because it only works with Steam Link, besides the Steam Link not having good quality and being extremely unstable, I can't use Steam's VirtualHere on Artemis or Moonlight, besides that the VirtualHere license is not worth it because the VirtualHere license is linked to the hardware and the developers told me that if I format my cell phone I lose my license, I've thought about using the USB Network Gate but it's very expensive and has no reverse connection, besides that since I exclusively use Moonlight and Artemis I would like the redirection of USB devices to be done directly from Artemis without the need to exit or minimize Artemis.

**Reply by ClassicOldSong** [reply]
As far as I know it also works with SteamLink on Android. There's a weird way to start VirtualHere with SteamLink Android App just to grant the license, then use other software for streaming, but it's in Chinese you have to use a translator to read: https://post.smzdm.com/p/awzd5dop/

**Reply by dsanttos12** [reply]
There were a lot of people commenting that it didn't work.
